Boto ec2 secgroup

github-abcde opened the pull request:
### What does this PR do?
It adds the states secgroup_present, secgroup_absent, and secgroup_equals to boto_ec2.
These allow you to add, remove or set the security-groups of already existing/running EC2 instances.
Cleans up sources of states/boto_ec2.py and modules/boto_ec2.py based on pylint feedback.

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/38580

### Previous Behavior
Securitygroups can only be specified upon creating an EC2-instance.

### New Behavior
Securitygroups can be modified for existing (and running) EC2 instances.

### Tests written?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

Commit:
states/boto_ec2.py, modules/boto_ec2.py: Added secgroup_present, secgroup_absent, secgroup_equals states that modify the security groups of an existing EC2 instance.

Commit:
states/boto_ec2.py: Added code-block examples

Commit:
salt/modules/boto_ec2.py: Pylint-inspired improvements.

Commit:
salt/states/boto_ec2.py: Pylint-inspired cleanup. Added return statement that went missing with rebase.

Commit:
salt/modules/boto_ec2.py: Renamed key variable to attrib, since there already was a key variable.

Commit:
salt/modules/boto_ec2.py: Updated docstring of describe_instances. Slight rewording of returned comment in modify_instances_secgroup.

Commit:
salt/modules/boto_ec2.py: Fix invalid input, if user does not supply a list of secgroup_names, wrap whatever was supplied in a list.

salt-jenkins commented:
@github-abcde, thanks for your PR! By analyzing the history of the files in this pull request, we identified @tkwilliams, @ryan-lane and @quantonganh to be potential reviewers.

ryan-lane commented on salt/states/boto_ec2.py:
```diff
@@ -1752,5 +1749,169 @@ def private_ips_absent(name, network_interface_name=None, network_interface_id=N
+    return ret
+
+
+def secgroup_present(
```
State modules are idempotent, so usually, rather than having present/equal/absent states for something you're going to apply to a resource, you'd add security groups as an attribute of the instance itself, and not manage attributes of a resource through several unrelated states.

Are you adding these because you have instances that are not managed by salt, but you want to manage parts of their resources? I'm not necessarily opposed to these states, since they're idempotent, but they're a bit out of consistency with how other things are done.

Does the instance_present module have a security groups attribute? If not, it would be ideal for this support to also be added there.

github-abcde replied:
I'm not quite sure I follow. Except for instance_present/absent, all other present/absent states in this file are managing resources of EC2 instances. Why would the security-group resources have to be any different?

The instance_present module does have a security_groups-attribute. However running that module with a different set of security-groups (or any resource for that matter) doesn't change the resource of the targeted EC2 instance. That is why issue #38580 exists.

github-abcde replied:
I could alter `secgroup_present` to behave like `secgroup_equals` by default (thereby being more in sync with other _present states) by adding a boolean option in the arguments. This would also remove the need to have a separate `secgroup_equals` defined.

ryan-lane replied:
It's inconsistent with the rest of the modules. The way the modules are meant to be designed is that everything related to a single resource is managed through a single state, so for instance, this is bad:

```
Ensure ec2 instance is created:
  boto_ec2.instance_present:
    - name: myec2instance

Ensure myec2instance has security groups:
  boto_ec2.secgroup_present:
    - name: myec2instance
    - security_groups:
       - secgroup1
       - secgroup2
```

This is good:

```
Ensure ec2 instance is created:
  boto_ec2.instance_present:
    - name: myec2instance
    - security_groups:
       - secgroup1
       - secgroup2
```

Folks have added extra functions in here to manage resources that aren't fully salt managed, which isn't ideal, but we accept them because sometimes folks need these as part of a transition.

It's fine to have these functions, as long as they aren't the primary way of maintaining parts of an ec2 instance.

github-abcde replied:
A use-case I'm encountering is that I want to change the security-groups after the instance has already been created (by salt). For instance, the default salt bootstrap requires access to Internet, but some EC2 instances should not have Internet access at all. In order to fix this, I create the instance with the relevant securitygroup, but update the security-groups in a later stage to not have the security-group that provides Internet Access. A later stage does not need to have (nor has) all the other EC2 instance configuration data, so a full `boto_ec2.instance_present` with all the same attributes as the original creation call, except the Internet access security group, is not feasible.

Another use-case is where I temporarily want to have a security-group present during a state run for temporary access from the Internet, and remove that security-group again later. You really don't want to spam the entire `boto_ec2.instance_present` with all its creation parameters twice in a state file (at least, I don't) just for a temporary security-group modification.

ryan-lane replied:
Yeah, that's totally reasonable, and why we have other helper functions like this, but it's also best to have the state properly manage its subresources as well. So, if it's not difficult, it would be ideal for changes to the secgroups in the instance_present function also idempotently maintain the instances secgroups.

ryan-lane replied:
I'm just saying we don't want the helpers to be the primary mechanism for managing the groups.

ryan-lane commented on salt/states/boto_ec2.py:
```diff
@@ -1752,5 +1749,169 @@ def private_ips_absent(name, network_interface_name=None, network_interface_id=N
+    Multifunction wrapper function for secgroup_present, secgroup_absent and secgroup_equals
+    '''
+    ret = {'name': name, 'result': True, 'changes': {}, 'comment': ''}
+    result = __salt__['boto_ec2.modify_instances_secgroup'](
```
You need to use test mode, or salt will make changes during the test run, which would be bad.

ryan-lane replied:
where by use I mean add support for test mode here.

github-abcde replied:
The test-mode is added to `salt/modules/boto_ec2.py:_modify_instance_secgroup`, which is the only place where modifications are actually made. The results of that will traverse back here.

ryan-lane replied:
That's really inconsistent with the rest of the modules and it makes it incredibly difficult to properly code review. The way salt is designed, execution modules are not idempotent, but state modules are. When I review the modules, I only check the state modules, based on this, and I expect test mode to be checked in them.

github-abcde replied:
Ok, I'll move the code back to the state module, but that will cause some code duplication, since each state function will have to fetch the current state, determine the differences and then check for testmode and actualize the changes (or not).

ryan-lane commented on salt/modules/boto_ec2.py:
```diff
@@ -2145,3 +2168,201 @@ def create_volume(zone_name, size=None, snapshot_id=None, volume_type=None,
+        AWS EC2 Instance object
+
+    The other arguments are the same as for modify_instances_secgroup
+    '''
```
Test mode

github-abcde replied:
Added Test-mode in a206240

Commit:
salt/modules/boto_ec2.py:
  modify_instance_secgroup:
    Added argument explaining comments
    Added ability to specify instance ID instead of nam
    Added False result if one of the instances modification returns False
    Replaced calls to __salt__[boto_ec2.describe_instances] with a direct call to describe_instances
  _modify_instance_secgroup:
    Added support for test-mode
    Bugfix in returning result:result:False because boto_ec2.modify_network_interface_attribute returns a dict
    Replaced calls to __salt__[boto_ec2.modify_network_interface_attribute] with a direct call to modify_network_interface_attribute.

cachedout commented:
@saltstack/team-boto Could y'all come by here and let me know if you think this is ready to go? Thanks.

github-abcde commented:
I still need to apply ryan-lane's directions to the code. Haven't had much time to work on it the past few weeks.

cachedout commented:
@github-abcde Thanks. Please keep us posted.

tkwilliams commented:
I would have to agree entirely with @ryan-lane here WRT the *_present and *_absent function naming.  'secgroup_present()' implies a function which ensures a security group with the given characteristics EXISTS as an EC2 resource.  Ensuring the presence - and state of - an instance is (or should be) done by 'instance_exists()' (or 'instance_present()' -- there are a lot of examples of both to be found).

Since what are being managed here are the security groups associated with an instance, it necessarily follows that this would be subsumed into the instance state management function, rather than a standalone function which then manages only certain attributes of that instance while leaving others untouched.  This latter would lead down a very fraught and quickly snowballing path of a new state function for every ATTRIBUTE of every resource type, which I don't imagine we want to even think about exploring :)

The best approach is to have some utility functions in the exec module for those cases where such manual tuning of secgroups is needed, along with an enhancement to the current instance state functions to utilize these new exec functions to update instances on-the-fly, as needed by changing  state defs...

WRT the test mode being in the exec function, I'm probably more lenient about that approach than Ryan :) but it IS definitely an unusual enough approach to warrant a clear callout in the state code explaining why it's done this way, and why it is thus safe... TBH, it's prolly better to just pass through the test=True as an explicit param to the exec functions from within the state function - then you won't need to duplicate any check code (since you're just passing it on and letting the exec functions handle it) but at the same time, a reviewer (or later code updater) will see clearly in front of them that you ARE aware of and coping with the test=True case ....

github-abcde commented:
Moved this to WIP_boto_ec2_secgroup. Will provide a PR again once I got it in order with respect to above comments.

damon-atkins commented:
Sounds like if you remove a security group from the list e.g. security group2 then it's still their.  So the name is wrong security-groups: should be security-groups-ensure. And we need added a security-group-absent.  The other option is a flag which indicates remove any security group not listed.

github-abcde commented:
@damon-atkins I divided the state into 3 parts:
- secgroup_present(list of groups): Adds any of the supplied list that are not already present. Ignores other existing groups.
- secgroup_absent(list of groups): Removes any of the supplied lists that are present. Ignores other existing groups.
- secgroup_equals(list of groups): Adds and/or removes groups such that only the list of groups supplied are present, no more, no less.

I'm not sure about your comment about an incorrect name; the _ensure you mention in combination with the flag is functionally the same as _present and _equals.

damon-atkins commented:
E.g. or a better name
```
Ensure ec2 instance is created:
  boto_ec2.instance_present:
    - name: myec2instance
    - security_groups_sync: True
    - security_groups:
       - secgroup1
       - secgroup2
```

anitakrueger commented:
@github-abcde Would you mind sharing what the latest status on this PR is please? I have the exact same use-case...instance created with Salt, but Salt needs ssh access for bootstrapping so I have that in a separate security group that needs to be removed from the instance later.
If you have some code already, I'd be happy to contribute as well.

damon-atkins commented:
For bootstrap you can always use AWS userdata to install a salt-minion. I am surprised this is not salts preferred method on aws.

github-abcde commented:
@anitakrueger Unfortunately, I haven't had the time to continue working on this. It is still on my to-do, but progress is slow and with long intervals.

anitakrueger commented:
@github-abcde I understand. Do you have any code you could share so someone else can work on it maybe?

@damon-atkins Would you mind sharing your method of installing the salt-minion? I haven't made much user of AWS userdata at all :(

damon-atkins commented:
https://github.com/number5/cloud-init/blob/master/doc/examples/cloud-config-salt-minion.txt
And many other examples.  And AWS doco.
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/user-data.html

github-abcde commented:
@anitakrueger The code I have is the code present in this PR. See the commits or files changed at the top of this page.
